File: modules/home/macos/cursor-warp/cursor-warp.m
#import <AppKit/AppKit.h>
#import <CoreGraphics/CoreGraphics.h>

#define WARP_WINDOW_MS 0.5

static CGEventRef eventTapCallback(CGEventTapProxy proxy, CGEventType type,
                                   CGEventRef event, void *refcon) {
  if (type == kCGEventTapDisabledByTimeout) {
    fprintf(stderr, "cursor-warp: event tap disabled by timeout\n");
    return NULL;
  }
  if (type != kCGEventKeyDown)
    return event;

  CGEventFlags flags = CGEventGetFlags(event);
  if (!(flags & kCGEventFlagMaskAlternate))
    return event;

  uint16_t key = CGEventGetIntegerValueField(event, kCGKeyboardEventKeycode);
  BOOL isFocusDir = (key == 4 || key == 38 || key == 40 || key == 37);
  BOOL isMoveToWS = (key >= 18 && key <= 29) && (flags & kCGEventFlagMaskShift);

  if (isFocusDir || isMoveToWS) {
    fprintf(stderr, "cursor-warp: scheduling warp\n");
    [(__bridge id)refcon performSelector:@selector(scheduleWarp)];
  }
  return event;
}

static void axObserverCallback(AXObserverRef observer, AXUIElementRef element,
                               CFStringRef notification, void *refcon) {
  [(__bridge id)refcon performSelector:@selector(onFocusChanged)];
}

@interface AppDelegate : NSObject <NSApplicationDelegate>
@property BOOL pendingWarp;
@property(strong) NSTimer *warpTimer;
@end

@implementation AppDelegate

- (void)applicationDidFinishLaunching:(NSNotification *)note {
  fprintf(stderr, "cursor-warp: started\n");

  if (!AXIsProcessTrusted()) {
    fprintf(stderr, "cursor-warp: accessibility permission not granted\n");
    return;
  }
  fprintf(stderr, "cursor-warp: accessibility granted\n");
  [self setupEventTap];
  [self setupAXObserver];

  dispatch_after(
      dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.3 * NSEC_PER_SEC)),
      dispatch_get_main_queue(), ^{
        fprintf(stderr, "cursor-warp: initial warp\n");
        [self warpToFocusedWindow];
      });
}

- (void)scheduleWarp {
  self.pendingWarp = YES;
  [self.warpTimer invalidate];
  self.warpTimer = [NSTimer scheduledTimerWithTimeInterval:WARP_WINDOW_MS
                                                   repeats:NO
                                                     block:^(NSTimer *t) {
                                                       self.pendingWarp = NO;
                                                     }];
}

- (void)onFocusChanged {
  if (!self.pendingWarp)
    return;
  self.pendingWarp = NO;
  [self.warpTimer invalidate];
  [self warpToFocusedWindow];
}

- (void)setupEventTap {
  CGEventMask mask = CGEventMaskBit(kCGEventKeyDown);
  CFMachPortRef tap = CGEventTapCreate(
      kCGSessionEventTap, kCGHeadInsertEventTap, kCGEventTapOptionDefault, mask,
      eventTapCallback, (__bridge void *)self);
  if (!tap || !CFMachPortIsValid(tap)) {
    NSLog(@"CursorWarp: failed to create event tap");
    return;
  }
  CFRunLoopAddSource(CFRunLoopGetCurrent(),
                     CFMachPortCreateRunLoopSource(kCFAllocatorDefault, tap, 0),
                     kCFRunLoopDefaultMode);
}

- (void)setupAXObserver {
  AXObserverRef observer;
  if (AXObserverCreate(getpid(), axObserverCallback, &observer) !=
      kAXErrorSuccess) {
    NSLog(@"CursorWarp: failed to create AX observer");
    return;
  }

  AXUIElementRef systemWide = AXUIElementCreateSystemWide();
  AXError err = AXObserverAddNotification(observer, systemWide,
                                          CFSTR("AXFocusedWindowChanged"),
                                          (__bridge void *)self);
  CFRelease(systemWide);

  if (err != kAXErrorSuccess) {
    NSLog(@"CursorWarp: AXObserverAddNotification failed (%d)", err);
    CFRelease(observer);
    return;
  }

  CFRunLoopAddSource(CFRunLoopGetCurrent(),
                     AXObserverGetRunLoopSource(observer),
                     kCFRunLoopDefaultMode);
}

- (void)warpToFocusedWindow {
  NSRunningApplication *frontApp =
      NSWorkspace.sharedWorkspace.frontmostApplication;
  if (!frontApp)
    return;

  AXUIElementRef appElement =
      AXUIElementCreateApplication(frontApp.processIdentifier);
  CFTypeRef focusedWindow;
  AXError err = AXUIElementCopyAttributeValue(
      appElement, CFSTR("AXFocusedWindow"), &focusedWindow);
  CFRelease(appElement);
  if (err != kAXErrorSuccess)
    return;

  AXUIElementRef window = focusedWindow;

  CFTypeRef positionRef;
  CFTypeRef sizeRef;
  AXUIElementCopyAttributeValue(window, CFSTR("AXPosition"), &positionRef);
  AXUIElementCopyAttributeValue(window, CFSTR("AXSize"), &sizeRef);
  CFRelease(window);

  if (!positionRef || !sizeRef) {
    if (positionRef)
      CFRelease(positionRef);
    if (sizeRef)
      CFRelease(sizeRef);
    return;
  }

  CGPoint position;
  CGSize size;
  Boolean ok = AXValueGetValue(positionRef, kAXValueCGPointType, &position);
  if (!ok) {
    CFRelease(positionRef);
    CFRelease(sizeRef);
    return;
  }
  ok = AXValueGetValue(sizeRef, kAXValueCGSizeType, &size);
  if (!ok) {
    CFRelease(positionRef);
    CFRelease(sizeRef);
    return;
  }
  CFRelease(positionRef);
  CFRelease(sizeRef);

  CGPoint center =
      CGPointMake(position.x + size.width / 2, position.y + size.height / 2);

  CGWarpMouseCursorPosition(center);
  CGEventRef moveEvent = CGEventCreateMouseEvent(NULL, kCGEventMouseMoved,
                                                 center, kCGMouseButtonLeft);
  if (moveEvent) {
    CGEventPost(kCGHIDEventTap, moveEvent);
    CFRelease(moveEvent);
  }
}

@end

int main(int argc, const char *argv[]) {
  @autoreleasepool {
    NSApplication *app = [NSApplication sharedApplication];
    AppDelegate *delegate = [[AppDelegate alloc] init];
    app.delegate = delegate;
    [app setActivationPolicy:NSApplicationActivationPolicyProhibited];
    [app run];
  }
  return 0;
}

File: modules/home/macos/cursor-warp/CursorWarp.swift
import AppKit
import Foundation

let warpWindowMs: TimeInterval = 0.5

class AppDelegate: NSObject, NSApplicationDelegate {
    var pendingWarp = false
    var warpTimer: Timer?

    // ── App Delegate ──────────────────────────────────────────────────────────

    func applicationDidFinishLaunching(_: Notification) {
        guard AXIsProcessTrusted() else {
            NSLog("CursorWarp: accessibility permission not granted")
            return
        }

        setupEventTap()
        setupAXObserver()

        // Warp once on startup to initial focused window
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
            self.warpToFocusedWindow()
        }
    }

    // ── Event Tap: detect AeroSpace focus-change keys ────────────────────────
    // AeroSpace keycodes (ANSI physical positions):
    //   h=4, j=38, k=40, l=37  (focus/move direction)
    //   numbers 1-0 = 18-29    (workspace select/move)

    let focusKeys: Set<UInt16> = [4, 38, 40, 37]

    func setupEventTap() {
        let eventMask = (1 << CGEventType.keyDown.rawValue)
        guard let tap = CGEvent.tapCreate(
            tap: .cgSessionEventTap,
            place: .headInsertEventTap,
            options: .defaultTap,
            eventsOfInterest: CGEventMask(eventMask),
            callback: { _, _, event, refcon in
                guard let refcon else { return Unmanaged.passUnretained(event) }
                let delegate = Unmanaged<AppDelegate>.fromOpaque(refcon).takeUnretainedValue()
                let opts = event.flags
                guard opts.contains(.maskAlternate) else { return Unmanaged.passUnretained(event) }
                let key = event.keyCode
                let isFocusDir = delegate.focusKeys.contains(key)
                let isMoveToWS = key >= 18 && key <= 29 && opts.contains(.maskShift)
                if isFocusDir || isMoveToWS {
                    delegate.pendingWarp = true
                    delegate.warpTimer?.invalidate()
                    delegate.warpTimer = Timer.scheduledTimer(
                        withTimeInterval: warpWindowMs,
                        repeats: false
                    ) { _ in delegate.pendingWarp = false }
                }
                return Unmanaged.passUnretained(event)
            },
            userInfo: Unmanaged.passUnretained(self).toOpaque()
        ), tap.isValid else {
            NSLog("CursorWarp: failed to create event tap")
            return
        }

        CFRunLoopAddSource(
            CFRunLoopGetCurrent(),
            CFMachPortCreateRunLoopSource(kCFAllocatorDefault, tap, 0),
            .defaultMode
        )
    }

    // ── AX Observer: detect focused window changes ──────────────────────────

    func setupAXObserver() {
        let systemWide = AXUIElementCreateSystemWide()
        var observer: AXObserver?
        let selfPtr = Unmanaged.passUnretained(self).toOpaque()
        let callback: AXObserverCallback = { _, _, _, refcon in
            guard let refcon else { return }
            let delegate = Unmanaged<AppDelegate>.fromOpaque(refcon).takeUnretainedValue()
            DispatchQueue.main.async {
                guard delegate.pendingWarp else { return }
                delegate.pendingWarp = false
                delegate.warpTimer?.invalidate()
                delegate.warpToFocusedWindow()
            }
        }
        guard AXObserverCreate(getpid(), callback, &observer) == .success,
              let observer
        else {
            NSLog("CursorWarp: failed to create AX observer")
            return
        }
        let result = AXObserverAddNotification(
            observer,
            systemWide,
            "AXFocusedWindowChanged" as CFString,
            selfPtr
        )
        if result != .success {
            NSLog("CursorWarp: AXObserverAddNotification failed (\(result.rawValue))")
            return
        }
        CFRunLoopAddSource(
            CFRunLoopGetCurrent(),
            AXObserverGetRunLoopSource(observer),
            .defaultMode
        )
    }

    // ── Warp ─────────────────────────────────────────────────────────────────

    func warpToFocusedWindow() {
        guard let frontApp = NSWorkspace.shared.frontmostApplication else { return }
        let appElement = AXUIElementCreateApplication(frontApp.processIdentifier)
        var focusedWindow: CFTypeRef?
        guard AXUIElementCopyAttributeValue(
            appElement, "AXFocusedWindow" as CFString, &focusedWindow
        ) == .success,
            let windowElement = focusedWindow as! AXUIElement?
        else { return }

        var positionRef: CFTypeRef?
        var sizeRef: CFTypeRef?
        AXUIElementCopyAttributeValue(windowElement, "AXPosition" as CFString, &positionRef)
        AXUIElementCopyAttributeValue(windowElement, "AXSize" as CFString, &sizeRef)
        guard
            let posValue = positionRef,
            let sizeValue = sizeRef,
            AXValueGetType(posValue as! AXValue) == .cgPoint,
            AXValueGetType(sizeValue as! AXValue) == .cgSize
        else { return }

        var position = CGPoint.zero
        var size = CGSize.zero
        AXValueGetValue(posValue as! AXValue, .cgPoint, &position)
        AXValueGetValue(sizeValue as! AXValue, .cgSize, &size)

        let center = CGPoint(
            x: position.x + size.width / 2,
            y: position.y + size.height / 2
        )
        CGWarpMouseCursorPosition(center)
        if let moveEvent = CGEvent(
            mouseEventSource: nil,
            mouseType: .mouseMoved,
            mouseCursorPosition: center,
            mouseButton: .left
        ) {
            moveEvent.post(tap: .cghidEventTap)
        }
    }
}

let delegate = AppDelegate()
NSApplication.shared.delegate = delegate
NSApplication.shared.setActivationPolicy(.prohibited)
NSApplication.shared.run()
